README.md
# time_overlapping

TimeOverlapping package is a lightweight and blazing fast for checking the time overlapping in the given DateTime ranges.

[![codecov](https://codecov.io/gh/MohanSaiManthri/time_overlapping/branch/main/graph/badge.svg?token=CHD4N0WLZM)](https://codecov.io/gh/MohanSaiManthri/time_overlapping)
[![Dart](https://github.com/MohanSaiManthri/time_overlapping/actions/workflows/dart.yml/badge.svg)](https://github.com/MohanSaiManthri/time_overlapping/actions/workflows/dart.yml)
[![pub.dev](https://img.shields.io/pub/v/time_overlapping?label=pub.dev&logo=data%3Aimage%2Fpng%3Bbase64%2CiVBORw0KGgoAAAANSUhEUgAAAA4AAAAOCAQAAAC1QeVaAAAABGdBTUEAALGPC%2FxhBQAAACBjSFJNAAB6JgAAgIQAAPoAAACA6AAAdTAAAOpgAAA6mAAAF3CculE8AAAAAmJLR0QA%2F4ePzL8AAAAHdElNRQflCAwFLhzpNkNgAAAAAW9yTlQBz6J3mgAAATFJREFUGNNN0M0qhGEYh%2FHrft73NaZoImNKkYUFCxt7KdYWFiymKKfgCBTHYMshWFkppVijJp81skBMM9P4SN557r8FC9f2t7sM4IN3SqhgUwxYSxe4KAIBvskoo7JN63XryN7C1JdF2oB1KOAhmWMoXNkHhuKkXxbqOQETEBdsyWatjAMo93PbC%2FsRuuQVP1bVV9XWX37vj7Gaw435tlzLwtfV%2BsMnb6rha3QX9SrXivjHn16X4kGweYYQgGO7bNAGiji5EbSjQwIShsOufrlPHSPZbPqJVfyhXft0A5Izvdic9dPUc7KJtb4PvJykaZ9lMRLTM70wQ4%2FdWYNBclTxse510ks%2FkQ5ZthbeODKoMUEXjYSRUEtjXlSJHhvuPY3Y7%2FgOJTSajofoCDI92m3ODyeBsv3LZ4DoAAAAJXRFWHRkYXRlOmNyZWF0ZQAyMDIxLTA4LTEyVDA1OjQ2OjI4KzAwOjAwwzEIWwAAACV0RVh0ZGF0ZTptb2RpZnkAMjAyMS0wOC0xMlQwNTo0NjoyOCswMDowMLJssOcAAAAASUVORK5CYII%3D&logoColor=green)](https://pub.dartlang.org/packages/time_overlapping)
[![license](https://img.shields.io/github/license/MohanSaiManthri/time_overlapping?logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQEAQAAADlauupAAAABGdBTUEAALGPC/xhBQAAACBjSFJNAAB6JgAAgIQAAPoAAACA6AAAdTAAAOpgAAA6mAAAF3CculE8AAAACXBIWXMAAABgAAAAYADwa0LPAAAAB3RJTUUH5QgMBTIvsJF/KwAAAAJiS0dEAACqjSMyAAAA6klEQVQ4y53TrwrCUBTH8cWNJatgVJF1weCf4DuIQebqlowKewqD2YcQ38GwB9jAIJoMrk1Qzle2C4rN469/zrnncK5l/RGwbeh0rP/wcAhZBiLqIhBF8HxSJU3LlyhwHIPvw+0GjwfS7yvwaoXMZnA8QhBAGP6OK1gmzw2OY0XnXg+KgneWSwWu1+F8NvB+h/lcgR0HDofKyvWKDAaKmVstWCwMThKk0VB0nk7N0jYbU8R1f4TtNux2sN3Cel1tW5pNxbyn02fTZZHJRDFzt8tX9nvlfY9GH3y5gOcpC9RqyHhsjsZxtL/zBX3z7zeyXyFwAAAAJXRFWHRkYXRlOmNyZWF0ZQAyMDIxLTA4LTEyVDA1OjUwOjQ3KzAwOjAwP4besgAAACV0RVh0ZGF0ZTptb2RpZnkAMjAyMS0wOC0xMlQwNTo1MDo0NyswMDowME7bZg4AAAAASUVORK5CYII=)](https://github.com/MohanSaiManthri/time_overlapping/blob/main/LICENSE)

## Getting Started

This package accepts only `List<OverlapParams>` 

#### OverlapParams
This requires an uniqueId, DateTimeRange & dynamic data*.
> dynamic data is type **dynamic** which accepts any sort of data.

It will throw **AssertionError** if it founds any duplicate Ids.
```dart
OverlapParams('uniqueId', DateTimeRange(start, end), {}),
```

There are 3 type of functions to find overlapping.
1. **findOverlap**  - returns the `List<String>` ie.., list of overlapping Ids.
2. **findOverlapWithData** - returns the `List<OverlapParams>` which has DateTimeRange overlap.
3. **hasOverlap** - to find timeOverlap for only 2 dateTimeRange. - returns bool.

## Usage
### findOverlap

```dart
void main() {
  final range1 = DateTimeRange(
    start: DateTime.now(),
    end: DateTime.now().add(const Duration(hours: 2)),
  );

  final range2 = DateTimeRange(
    start: DateTime.now(),
    end: DateTime.now().add(const Duration(hours: 1)),
  );

  final range3 = DateTimeRange(
    start: DateTime.now().add(const Duration(hours: 2)),
    end: DateTime.now().add(const Duration(hours: 3)),
  );

  final range4 = DateTimeRange(
    start: DateTime.now().add(const Duration(hours: 3)),
    end: DateTime.now().add(const Duration(hours: 4)),
  );

  final overlapParams = [
    OverlapParams('001', range1, {}),
    OverlapParams('002', range2, {}),
    OverlapParams('003', range3, {}),
    OverlapParams('004', range4, {}),
  ];

  print(TimeOverlapFinder.findOverlap(overlapParams));
}
```
##### Output
```dart
['001', '002']
```

### findOverlapWithData

```dart
void main() {
  final range1 = DateTimeRange(
    start: DateTime.now(),
    end: DateTime.now().add(const Duration(hours: 2)),
  );

  final range2 = DateTimeRange(
    start: DateTime.now(),
    end: DateTime.now().add(const Duration(hours: 1)),
  );

  final range3 = DateTimeRange(
    start: DateTime.now().add(const Duration(hours: 2)),
    end: DateTime.now().add(const Duration(hours: 3)),
  );

  final range4 = DateTimeRange(
    start: DateTime.now().add(const Duration(hours: 3)),
    end: DateTime.now().add(const Duration(hours: 4)),
  );

  final overlapParams = [
    OverlapParams('001', range1, {}),
    OverlapParams('002', range2, {}),
    OverlapParams('003', range3, {}),
    OverlapParams('004', range4, {}),
  ];

  print(TimeOverlapFinder.findOverlapWithData(overlapParams));
}
```
##### Output
```dart
 [
    OverlapParams(001, 2021-08-12 07:36:07.095 - 2021-08-12 09:36:07.095, {}),
    OverlapParams(002, 2021-08-12 07:36:07.095 - 2021-08-12 08:36:07.095, {})
 ]
```

### hasOverlap

```dart
void main() {
  final range1 = DateTimeRange(
    start: DateTime.now(),
    end: DateTime.now().add(const Duration(hours: 2)),
  );

  final range2 = DateTimeRange(
    start: DateTime.now(),
    end: DateTime.now().add(const Duration(hours: 1)),
  );

  final overlapParams = [
    OverlapParams('001', range1, {}),
    OverlapParams('002', range2, {}),
  ];

  print(TimeOverlapFinder.hasOverlap(range1, range2));
}
```

##### Output
```dart
true
```

#### Some useful extensions

```dart
1. range1.isBefore(range2),
2. range1.isAfter(range2),
```

```dart
1. range1.isCrossing(range2),
2. range1.isCrossingStart(range2),
3. range1.isCrossingEnd(range2),
```

### Thanks for using my package 🙏


CHANGELOG.md
## 0.1.0

* Minimum version of Flutter (3.3.10 now)
* Added few more tests to check allowTouches and ignoreSeconds

## 0.0.5

* `allowTouches`, `ignoreSeconds` functionality added [#1] Github issue - Resolved.

## 0.0.4

* Code formatting.

## 0.0.3

* `hasOverlap` method fix.
* Core functionality improved.

## 0.0.2

* Initial release.


test/time_overlapping_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:time_overlapping/time_overlapping.dart';

void main() {
  group('Overlap Tests: ', () {
    test('Testing duplicate Id\'s', () {
      final range1 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 2)),
      );

      final range2 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 1)),
      );

      final range3 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 2)),
        end: DateTime.now().add(const Duration(hours: 3)),
      );

      final range4 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 3)),
        end: DateTime.now().add(const Duration(hours: 4)),
      );

      final overlapParams = [
        OverlapParams('001', range1, {}),
        OverlapParams('002', range2, {}),
        OverlapParams('001', range3, {}),
        OverlapParams('004', range4, {}),
      ];

      expect(() => TimeOverlapFinder.findOverlap(overlapParams), throwsAssertionError);
      expect(() => TimeOverlapFinder.findOverlapWithData(overlapParams), throwsAssertionError);
    });

    test('Testing ordered overlap ranges', () {
      final range1 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 2)),
      );

      final range2 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 1)),
      );

      final range3 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 2)),
        end: DateTime.now().add(const Duration(hours: 3)),
      );

      final range4 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 3)),
        end: DateTime.now().add(const Duration(hours: 4)),
      );

      final overlapParams = [
        OverlapParams('001', range1, {}),
        OverlapParams('002', range2, {}),
        OverlapParams('003', range3, {}),
        OverlapParams('004', range4, {}),
      ];
      expect(TimeOverlapFinder.findOverlap(overlapParams), ['001', '002']);
    });

    test('Testing ordered overlap ranges: 2', () {
      final range1 = DateTimeRange(
        start: DateTime(2021, 8, 12, 10, 00),
        end: DateTime(2021, 8, 12, 10, 10),
      );

      final range2 = DateTimeRange(
        start: DateTime(2021, 8, 12, 10, 00),
        end: DateTime(2021, 8, 12, 10, 20),
      );

      final range3 = DateTimeRange(
        start: DateTime(2021, 8, 12, 10, 00),
        end: DateTime(2021, 8, 12, 10, 15),
      );

      final overlapParams = [
        OverlapParams('001', range1, {}),
        OverlapParams('002', range2, {}),
        OverlapParams('003', range3, {}),
      ];
      expect(TimeOverlapFinder.findOverlap(overlapParams), ['001', '002', '003']);
    });
    test('Testing unOrdered overlap ranges', () {
      final range1 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 2)),
      );

      final range2 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 2)),
        end: DateTime.now().add(const Duration(hours: 3)),
      );

      final range3 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 1)),
      );

      final range4 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 3)),
        end: DateTime.now().add(const Duration(hours: 4)),
      );

      final overlapParams = [
        OverlapParams('001', range1, {}),
        OverlapParams('002', range2, {}),
        OverlapParams('003', range3, {}),
        OverlapParams('004', range4, {}),
      ];
      expect(TimeOverlapFinder.findOverlap(overlapParams), ['001', '003']);
    });

    test('Testing unOrdered overlap ranges with data', () {
      final range1 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 2)),
      );

      final range2 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 2)),
        end: DateTime.now().add(const Duration(hours: 3)),
      );

      final range3 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 1)),
      );

      final range4 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 3)),
        end: DateTime.now().add(const Duration(hours: 4)),
      );

      final overlapParams = [
        OverlapParams('001', range1, {'name': 'userA'}),
        OverlapParams('002', range2, {'name': 'userB'}),
        OverlapParams('003', range3, {'name': 'userC'}),
        OverlapParams('004', range4, {'name': 'userD'}),
      ];
      expect(
        TimeOverlapFinder.findOverlapWithData(overlapParams),
        [
          OverlapParams('001', range1, {'name': 'userA'}),
          OverlapParams('003', range3, {'name': 'userC'}),
        ],
      );
    });

    test('Testing time complexity for long range', () {
      final range1 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 2)),
      );

      final range2 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 2)),
        end: DateTime.now().add(const Duration(hours: 3)),
      );

      final range3 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 1)),
      );

      final range4 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 3)),
        end: DateTime.now().add(const Duration(hours: 4)),
      );

      final range5 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 3)),
        end: DateTime.now().add(const Duration(hours: 5)),
      );

      final range6 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 3)),
        end: DateTime.now().add(const Duration(hours: 3, minutes: 30)),
      );

      final range7 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 5)),
        end: DateTime.now().add(const Duration(hours: 6)),
      );

      final range8 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 6)),
        end: DateTime.now().add(const Duration(hours: 9)),
      );

      final range9 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 6)),
        end: DateTime.now().add(const Duration(hours: 8)),
      );

      final range10 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 7)),
        end: DateTime.now().add(const Duration(hours: 8)),
      );

      final range11 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 9)),
        end: DateTime.now().add(const Duration(hours: 10)),
      );

      final range12 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 8)),
        end: DateTime.now().add(const Duration(hours: 10)),
      );

      final range13 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 10)),
        end: DateTime.now().add(const Duration(hours: 11)),
      );

      final range14 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 11)),
        end: DateTime.now().add(const Duration(hours: 12)),
      );

      final range15 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 12)),
        end: DateTime.now().add(const Duration(hours: 13)),
      );

      final overlapParams = [
        OverlapParams('001', range1, {}),
        OverlapParams('002', range2, {}),
        OverlapParams('003', range3, {}),
        OverlapParams('004', range4, {}),
        OverlapParams('005', range5, {}),
        OverlapParams('006', range6, {}),
        OverlapParams('007', range7, {}),
        OverlapParams('008', range8, {}),
        OverlapParams('009', range9, {}),
        OverlapParams('0010', range10, {}),
        OverlapParams('0011', range11, {}),
        OverlapParams('0012', range12, {}),
        OverlapParams('0013', range13, {}),
        OverlapParams('0014', range14, {}),
        OverlapParams('0015', range15, {}),
      ];
      expect(TimeOverlapFinder.findOverlap(overlapParams), [
        '001',
        '003',
        '004',
        '005',
        '006',
        '008',
        '009',
        '0010',
        '0012',
        '0011',
      ]);
    });

    test('Testing single dateTimeRange', () {
      final range1 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 2)),
      );

      final overlapParams = [
        OverlapParams('001', range1, {}),
      ];

      expect(TimeOverlapFinder.findOverlap(overlapParams), <String>[]);
    });

    test('Testing single dateTimeRange with data', () {
      final range1 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 2)),
      );

      final overlapParams = [
        OverlapParams('001', range1, {'name': 'userA'}),
      ];

      expect(TimeOverlapFinder.findOverlapWithData(overlapParams), <String>[]);
    });

    test('Testing hasOverlap method: 1', () {
      final range1 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 2)),
      );

      final range2 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 2)),
        end: DateTime.now().add(const Duration(hours: 3)),
      );
      expect(TimeOverlapFinder.hasOverlap(range1, range2), false);
    });

    test('Testing findOverlap  for 2 items without overlapping', () {
      final range1 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 2)),
      );

      final range2 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 2)),
        end: DateTime.now().add(const Duration(hours: 3)),
      );

      final overlapParams = [
        OverlapParams('001', range1, {'name': 'userA'}),
        OverlapParams('002', range2, {'name': 'userB'}),
      ];

      expect(TimeOverlapFinder.findOverlap(overlapParams), []);
    });

    test('Testing findOverlap  for 2 items with overlapping', () {
      final range1 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 2)),
      );

      final range2 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 1)),
      );

      final overlapParams = [
        OverlapParams('001', range1, {'name': 'userA'}),
        OverlapParams('002', range2, {'name': 'userB'}),
      ];

      expect(TimeOverlapFinder.findOverlap(overlapParams), ['001', '002']);
    });

    test('Testing hasOverlap method: 2', () {
      final range1 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 2)),
      );

      final range2 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 1)),
      );

      expect(TimeOverlapFinder.hasOverlap(range1, range2), true);
    });

    test('Testing Allow touches: Allow touches `False`', () {
      final _commonTime = DateTime.now().add(const Duration(hours: 2));
      final range1 = DateTimeRange(
        start: DateTime.now(),
        end: _commonTime,
      );

      final range2 = DateTimeRange(
        start: _commonTime,
        end: _commonTime.add(const Duration(hours: 1)),
      );
      print(range1);
      print(range2);
      expect(TimeOverlapFinder.hasOverlap(range1, range2, allowTouches: false), true);
    });

    test('Testing Allow touches: Allow touches `True`', () {
      final _commonTime = DateTime.now().add(const Duration(hours: 2));
      final range1 = DateTimeRange(
        start: DateTime.now(),
        end: _commonTime,
      );

      final range2 = DateTimeRange(
        start: _commonTime,
        end: _commonTime.add(const Duration(hours: 1)),
      );
      print(range1);
      print(range2);
      expect(TimeOverlapFinder.hasOverlap(range1, range2, allowTouches: true), false);
    });

    test('Testing Allow touches: Allow touches False Ignore Seconds: False', () {
      final _commonTime = DateTime.now().add(const Duration(hours: 2));
      final range1 = DateTimeRange(
        start: DateTime.now(),
        end: _commonTime,
      );

      final range2 = DateTimeRange(
        start: _commonTime,
        end: _commonTime.add(const Duration(hours: 1)),
      );
      print(range1);
      print(range2);
      expect(
          TimeOverlapFinder.hasOverlap(
            range1,
            range2,
            allowTouches: false,
            ignoreSeconds: false,
          ),
          true);
    });

    test('Testing Allow touches: Allow touches False Ignore Seconds: True', () {
      final range1 = DateTimeRange(start: DateTime(2023, 05, 18), end: DateTime(2023, 05, 18, 17));
      final range2 = DateTimeRange(
        start: DateTime(2023, 05, 18, 17, 0, 0, 299),
        end: DateTime(2023, 05, 18, 19, 0, 0, 299),
      );
      final range3 = DateTimeRange(start: DateTime(2023, 05, 18, 19), end: DateTime(2023, 05, 18, 23, 59));
      print(range1);
      print(range2);
      print(range3);
      expect(
        TimeOverlapFinder.hasOverlap(
          range1,
          range2,
          allowTouches: false,
          ignoreSeconds: true,
        ),
        true,
      );
    });

    test('Testing Allow touches: Allow touches True Ignore Seconds: True', () {
      final range1 = DateTimeRange(start: DateTime(2023, 05, 18), end: DateTime(2023, 05, 18, 17));
      final range2 = DateTimeRange(
        start: DateTime(2023, 05, 18, 17, 0, 0, 299),
        end: DateTime(2023, 05, 18, 19, 0, 0, 299),
      );
      final range3 = DateTimeRange(start: DateTime(2023, 05, 18, 19), end: DateTime(2023, 05, 18, 23, 59));
      print(range1);
      print(range2);
      print(range3);
      expect(
        TimeOverlapFinder.hasOverlap(
          range1,
          range2,
          allowTouches: true,
          ignoreSeconds: true,
        ),
        false,
      );
    });

    test('Testing isCrossing function', () {
      final range1 = DateTimeRange(
        start: DateTime.now(),
        end: DateTime.now().add(const Duration(hours: 2)),
      );

      final range2 = DateTimeRange(
        start: DateTime.now().add(const Duration(hours: 1, minutes: 30)),
        end: DateTime.now().add(const Duration(hours: 3)),
      );

      expect(range2.isCrossing(range1), true);
      expect(range1.isCrossing(range2), true);
      expect(range1.isCrossingStart(range2), true);
      expect(range2.isCrossingStart(range1), false);
      expect(range2.isCrossingEnd(range1), true);
      expect(range1.isCrossingEnd(range2), false);
    });
  });
}


lib/src/overlap.dart
import 'package:flutter/material.dart';
import 'package:time_overlapping/time_overlapping.dart';

class TimeOverlapFinder {
  const TimeOverlapFinder._internal();

  /// returns all the overlapping Id's ie.., `List<String>`.
  ///
  /// If data is required then use
  /// ```dart
  /// TimeOverlapFinder.findOverlapWithData(overlapParams);
  /// ```
  /// ...
  static List<String> findOverlap(
    List<OverlapParams> overlapParams, {
    bool allowTouches = true,
    bool ignoreSeconds = true,
  }) {
    try {
      if (overlapParams.length < 2) {
        return <String>[];
      }
      assert(
        _validEntries(overlapParams),
        "There should be exactly one item with [OverlapParams]'s id.\n"
        'Either zero or 2 or more [OverlapParams]s were detected '
        'with the same value',
      );
      return _calculateOverlapRanges<String>(overlapParams);
    } on AssertionError catch (_) {
      rethrow;
    }
  }

  /// returns `List<OverlapParams>`
  /// This can be used if data needs to returned, so that it can be used
  /// for some purpose.
  ///
  /// If data is not required then use
  /// ```dart
  /// TimeOverlapFinder.findOverlap(overlapParams);
  /// ```
  /// ...
  static List<OverlapParams> findOverlapWithData(
    List<OverlapParams> overlapParams, {
    bool allowTouches = true,
    bool ignoreSeconds = true,
  }) {
    try {
      if (overlapParams.length < 2) {
        return <OverlapParams>[];
      }
      assert(
        _validEntries(overlapParams),
        "There should be exactly one item with [OverlapParams]'s id.\n"
        'Either zero or 2 or more [OverlapParams]s were detected '
        'with the same value',
      );
      return _calculateOverlapRanges<OverlapParams>(overlapParams);
    } on AssertionError catch (_) {
      rethrow;
    }
  }

  /// `allowTouches` defaults to true thus ignore any start, end overlapping
  /// Eg:
  /// ```dart
  /// 2022-03-12 11:47 - 2022-03-12 13:47
  /// 2022-03-12 13:47 - 2022-03-12 14:47
  /// ```
  ///
  /// `allowTouches: true` will return hasOverlap false;
  /// `false` will return hasOverlap true;
  ///
  ///
  /// ```dart
  /// if ignoreSeconds: true
  /// ```
  /// will makes seconds to `0` thus comparing
  /// only date, hour and minutes.
  ///
  static bool hasOverlap(
    DateTimeRange range1,
    DateTimeRange range2, {
    bool allowTouches = true,
    bool ignoreSeconds = true,
  }) {
    bool _hasOverlap = !(range2.isBefore(range1) || range2.isAfter(range1));
    if (allowTouches) return _hasOverlap;
    return hasTouch(range1, range2, ignoreSeconds: ignoreSeconds);
  }

  /// Checks the range has any touches at start or end
  /// Default: Ranges start and end's milliseconds will make `0` before comparing
  /// ```dart
  /// if ignoreSeconds: true
  /// ```
  /// will makes seconds to `0` thus comparing
  /// only date, hour and minutes.
  ///
  static bool hasTouch(
    DateTimeRange range,
    DateTimeRange other, {
    bool ignoreSeconds = false,
  }) {
    final maskedRange =
        ignoreSeconds ? range.maskSeconds() : range.maskMilliSeconds();
    final maskedOther =
        ignoreSeconds ? other.maskSeconds() : other.maskMilliSeconds();
    return maskedRange.start.isAtSameMomentAs(maskedOther.start) ||
        maskedRange.end.isAtSameMomentAs(maskedOther.start) ||
        maskedRange.start.isAtSameMomentAs(maskedOther.end) ||
        maskedRange.end.isAtSameMomentAs(maskedOther.end);
  }

  static List<T> _calculateOverlapRanges<T>(
    List<OverlapParams> ranges, {
    bool allowTouches = true,
    bool ignoreSeconds = true,
  }) {
    final overlapList = <T>{};

    /// [sort] all the ranges, so that we can easily find the overlapping.
    /// instead of dangling between indexes.
    ranges
        .sort((a, b) => a.dateTimeRange.start.compareTo(b.dateTimeRange.start));

    if (ranges.length == 2) {
      final hasTimeOverlap = hasOverlap(
        ranges.first.dateTimeRange,
        ranges.last.dateTimeRange,
        allowTouches: allowTouches,
        ignoreSeconds: ignoreSeconds,
      );

      if (hasTimeOverlap) {
        if (T == String) {
          overlapList.add(ranges.first.id as T);
          overlapList.add(ranges.last.id as T);
          return overlapList.toList();
        } else if (T == OverlapParams) {
          return ranges as List<T>;
        }
      } else {
        return overlapList.toList();
      }
    }

    for (var i = 0; i < ranges.length - 1; i++) {
      /// [subIndex] is an integer which points to next index.
      var subIndex = i + 1;
      var rangeAtIndex = ranges.elementAt(i);
      var rangeAtFutureIndex = ranges.elementAt(i + 1);

      var isBefore =
          rangeAtIndex.dateTimeRange.isBefore(rangeAtFutureIndex.dateTimeRange);
      var isAfter =
          rangeAtIndex.dateTimeRange.isAfter(rangeAtFutureIndex.dateTimeRange);

      ///
      /// 1.
      /// |--- A ---|
      ///           | --- B ---|
      ///
      /// or
      ///
      /// 2.
      /// |--- B---|
      ///          |--- A ---|
      ///
      /// [1] & [2] are positive conditions and [NOT] of that is overlapping cond.
      ///
      ///
      /// This will iterate till the moment we dont have overlap with next
      /// dateTimeRange.
      while ((!(isBefore || isAfter)) && subIndex < ranges.length) {
        subIndex += 1;
        if (T == String) {
          overlapList.add(rangeAtIndex.id as T);
          overlapList.add(rangeAtFutureIndex.id as T);
        } else if (T == OverlapParams) {
          overlapList.add(rangeAtIndex as T);
          overlapList.add(rangeAtFutureIndex as T);
        }

        if (subIndex >= ranges.length - 1) {
          break;
        }

        rangeAtIndex = ranges.elementAt(i);
        rangeAtFutureIndex = ranges.elementAt(subIndex);

        isBefore = rangeAtIndex.dateTimeRange
            .isBefore(rangeAtFutureIndex.dateTimeRange);
        isAfter = rangeAtIndex.dateTimeRange
            .isAfter(rangeAtFutureIndex.dateTimeRange);
      }
    }
    return overlapList.toList();
  }

  static bool _validEntries(List<OverlapParams> overlapParams) {
    final listOfIds = overlapParams.fold<List<String>>([overlapParams.first.id],
        (previousValue, element) {
      if (previousValue.contains(element.id)) {
        return previousValue;
      } else {
        return [...previousValue, element.id];
      }
    });
    return listOfIds.length == overlapParams.length;
  }
}


lib/src/extensions/date_time_extension.dart
import 'package:flutter/material.dart';

extension DateTimeRangeExtensions on DateTimeRange {
  /// [isAfter] returns whether the timeRange is after other timeRange or not.
  //
  // |-- A --|
  //         |-- B --|
  //
  bool isAfter(DateTimeRange other) {
    final otherRangeStart = other.start;
    final pointInRange = otherRangeStart.isAtSameMomentAs(end);
    final isAfterRange = otherRangeStart.isAfter(end);
    final isNotBeforeRange = !otherRangeStart.isBefore(end);
    return (pointInRange || isAfterRange) && isNotBeforeRange;
  }

  /// [isBefore] returns whether the timeRange is before than other timeRange.
  //
  // |-- B --|
  //         |-- A --|
  //
  bool isBefore(DateTimeRange other) {
    final otherRangeEnd = other.end;
    final pointInRange = otherRangeEnd.isAtSameMomentAs(start);
    final isBeforeRange = otherRangeEnd.isBefore(start);
    final isNotAfterRange = !otherRangeEnd.isAfter(start);
    return (pointInRange || isBeforeRange) && isNotAfterRange;
  }

  /// [isCrossing] will help whether this timerange is crossing other or not.
  /// Either it can start or end. it will return bool.
  // 1.
  // |-- A --|
  //        |-- B --|
  // 2.
  //        |-- A --|
  // |-- B --|
  bool isCrossing(DateTimeRange other) {
    return isCrossingStart(other) || isCrossingEnd(other);
  }

  /// [isCrossingStart] will help whether this timeRange is crossing other timerange start or not.
  // Eg:
  //|-- A --|
  //       |-- B --|
  bool isCrossingStart(DateTimeRange other) {
    return end.isBefore(other.start) || start.isBefore(other.start);
  }

  /// [isCrossingEnd] will help whether this timeRange crossing other timerange end or not.
  // Eg:
  // |-- B --|
  //        |-- A --|
  bool isCrossingEnd(DateTimeRange other) {
    return start.isAfter(other.end) || end.isAfter(other.end);
  }

  /// Makes both start & end of a timeRange `seconds` `millisecond` `microsecond` to `0`
  DateTimeRange maskSeconds() {
    return DateTimeRange(
        start: DateTime(
          start.year,
          start.month,
          start.day,
          start.hour,
          start.minute,
        ),
        end: DateTime(
          end.year,
          end.month,
          end.day,
          end.hour,
          end.minute,
        ));
  }

  /// Makes both start & end of a timeRange  `millisecond` `microsecond` to `0`
  DateTimeRange maskMilliSeconds() {
    return DateTimeRange(
        start: DateTime(
          start.year,
          start.month,
          start.day,
          start.hour,
          start.minute,
          start.second,
        ),
        end: DateTime(
          end.year,
          end.month,
          end.day,
          end.hour,
          end.minute,
          end.second,
        ));
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:time_overlapping/time_overlapping.dart';

void main() {
  final range1 = DateTimeRange(
    start: DateTime.now(),
    end: DateTime.now().add(const Duration(hours: 2)),
  );

  final range2 = DateTimeRange(
    start: DateTime.now(),
    end: DateTime.now().add(const Duration(hours: 1)),
  );

  final range3 = DateTimeRange(
    start: DateTime.now().add(const Duration(hours: 2)),
    end: DateTime.now().add(const Duration(hours: 3)),
  );

  final range4 = DateTimeRange(
    start: DateTime.now().add(const Duration(hours: 3)),
    end: DateTime.now().add(const Duration(hours: 4)),
  );

  final overlapParams = [
    OverlapParams('001', range1, {}),
    OverlapParams('002', range2, {}),
    OverlapParams('003', range3, {}),
    OverlapParams('004', range4, {}),
  ];

  print(TimeOverlapFinder.findOverlap(overlapParams));
  print(TimeOverlapFinder.findOverlapWithData(overlapParams));
}
